shared/teams/emojis/add-emoji.tsx

```diff
@@ -0,0 +1,597 @@
+      <Kb.Text type="BodySmall">Maximum 256KB per image.</Kb.Text>
+    </Kb.Box2>
+  )
+}
+
+type AddEmojiAliasAndConfirmProps = {
+  addFiles: (filePaths: Array<string>) => void
+  emojisToAdd: Array<EmojiToAdd>
+}
+
+type EmojiToAdd = {
+  alias: string
+  error: string
+  onChangeAlias: (newAlias: string) => void
+  path: string
+}
+
+type EmojiToAddOrAddRow =
+  | {
+      type: 'emoji'
+      emojiToAdd: EmojiToAdd
+      height: number
+      key: string
+      offset: number
+    }
+  | {
+      type: 'add'
+      add: () => any
+      height: number
+      key: string
+      offset: number
+    }
+
+const renderRow = (_: number, item: EmojiToAddOrAddRow) =>
+  item.type === 'add' ? (
+    <Kb.Box2 direction="horizontal" alignItems="center" fullWidth={true} style={styles.emojiToAddRow}>
+      <Kb.ClickableBox onClick={item.add} style={styles.addEmojiIconContainer}>
+        <Kb.Icon type="iconfont-new" color={Styles.globalColors.blue} />
+      </Kb.ClickableBox>
+    </Kb.Box2>
+  ) : (
+    <Kb.Box2
+      direction="vertical"
+      fullWidth={true}
+      gap="xxtiny"
+      style={Styles.collapseStyles([
+        styles.emojiToAddRow,
+        item.emojiToAdd.error && styles.emojiToAddRowWithError,
+      ])}
+    >
+      <Kb.Box2 direction="horizontal" gap="xsmall" fullWidth={true}>
+        <Kb.Box style={styles.emojiToAddImageContainer}>
+          <Kb.Image src={item.emojiToAdd.path} style={styles.emojiToAddImage} />
+        </Kb.Box>
+        <Kb.NewInput
+          error={!!item.emojiToAdd.error}
+          textType={Styles.isMobile ? 'BodySemibold' : 'Body'}
+          value={`:${item.emojiToAdd.alias}:`}
+          containerStyle={styles.aliasInput}
+          onChangeText={newText => item.emojiToAdd.onChangeAlias(newText.replace(/:/g, ''))}
+        />
+      </Kb.Box2>
+      {!!item.emojiToAdd.error && (
+        <Kb.Text type="BodySmallError" style={styles.errorText}>
+          {item.emojiToAdd.error}
+        </Kb.Text>
+      )}
+    </Kb.Box2>
+  )
+
+const AddEmojiAliasAndConfirm = (props: AddEmojiAliasAndConfirmProps) => {
+  const {dragOver, onDragLeave, onDragOver, onDrop, pick} = usePickFiles(props.addFiles)
+  const {emojisToAdd} = props
+  const items = React.useMemo(() => {
+    const ret = emojisToAdd.reduce((arr, emojiToAdd, index) => {
+      const last = arr[index - 1]
+      arr.push({
+        emojiToAdd,
+        height: emojiToAdd.error ? emojiToAddRowHeightWithError : emojiToAddRowHeightNoError,
+        key: emojiToAdd.path,
+        offset: last ? last.offset + last.height : 0,
+        type: 'emoji',
+      })
+      return arr
+    }, [] as Array<EmojiToAddOrAddRow>)
+    const last = ret[ret.length - 1]
+    ret.push({
+      add: pick,
+      height: emojiToAddRowHeightNoError,
+      key: 'btn:add',
+      offset: last ? last.offset + last.height : 0,
+      type: 'add',
+    })
+    return ret
+  }, [emojisToAdd, pick])
+
+  const rowItemHeight = React.useMemo(() => {
+    // @ts-ignore
```

Comment:
this isn't a pattern we shouldn't really be doing

Reply:
What should we be doing here instead?